### PR TITLE
Basic implementation of the Apache Kafka based storage

### PR DESCRIPTION
**Pull request by dmdashenkov**
Introducing the **Apache Kafka**-based Spine storage.

The PR contains two new modules:
 - `kafka-storage` - the storage library;
 - `deployment/local-kafka` - the TodoList local server starter which relies on the Kafka-ased storage.

Currently, the storage implementation has several limitations. See `io.spine.server.storage.kafka.package-info.java` for more details.

For the detailed user guide see `kafka-storage/README.MD`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.
Note: this is still not the final version since some tests are still broken (due to the reasons connected with the tests, not with the production code).

**Comment by dmdashenkov** [comment]
@dmytro-grankin, PTAL.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java [hunk 1]
```diff
@@ -113,6 +123,21 @@ public static BoundedContext create(StorageFactory storageFactory) {
...
 
+    /**
+     * Injects the given {@link StorageFactory} implementation.
+     *
+     * <p>After calling this method, invocation of {@link #create()} creates create a
+     * {@code BoundedContext} with the passed {@code StorageFactory}.
+     *
+     * @param storageFactory the default {@code StorageFactory} to create the BoundedContexts for
+     */
+    public static void injectStorageFactory(StorageFactory storageFactory) {
```
Comment (dmytro-grankin) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: Let's discuss the purpose of this method if there is `BoundedContexts.create(StorageFactory)`.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        return FOR_AGGREGATE_RECORD.create(ofType);
+    }
+
+    /**
+     * Creates an instance of {@link Topic} for the records of entity of given type.
+     *
+     * @param ofType the type of record
+     * @return new instance of {@link Topic}
+     */
+    static Topic forLifecycleFlags(Class<?> ofType) {
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java]: Fixed.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaAggregateStorage.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    @Override
+    protected void writeRecord(I id, AggregateEventRecord record) {
+        checkNotClosed();
+        final Topic topic = Topic.forAggregateRecord(aggregateClass);
+        Object key = null;
+        if (record.getKindCase() == EVENT) {
+            key = record.getEvent().getId();
+        } else if (record.getKindCase() == SNAPSHOT) {
+            key = topic.getName();
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaAggregateStorage.java]: Yes, it does. The `Topic` type represents a logical Entity type and the snapshot is a singleton instance per type.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/MessageSerializer.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+     */
+    public static Serializer<Message> serializer() {
+        return Default.INSTANCE.value;
+    }
+
+    /**
+     * Returns an instance of {@code MessageSerializer} as a {@link Deserializer}.
+     */
+    public static Deserializer<Message> deserializer() {
+        return Default.INSTANCE.value;
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/storage/kafka/MessageSerializer.java]: When using this class, we do not need the whole interface at once, though the work of both interfaces boils down to a single responsibility - the `Message` serialization.

**Comment by dmdashenkov** [comment]
@dmytro-grankin, PTAL again.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStandStorage.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        return delegate.readRecord(key);
+    }
+
+    @Override
+    protected Iterator<EntityRecord> readMultipleRecords(Iterable<AggregateStateId> ids) {
+        @SuppressWarnings("StaticPseudoFunctionalStyleMethod")
+            // Better in several reasons:
+            //  1. Executed lazily.
+            //  2. Shorter notation.
+            //  3. Potentially less GC overhead.
```
Comment (dmytro-grankin) [inline kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStandStorage.java]: For such cases you can use the following format - `@SuppressWarnings("..." /* comment */)`. This prevents a shift of the comments after reformatting.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStorageFactory.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+     *
+     * <p>The value is equal to 1 second.
+     *
+     * @see Builder#setMaxPollAwait(Duration)
+     */
+    private static final Duration DEFAULT_MAX_POLL_AWAIT = Duration.of(1, SECONDS);
+
+    private final KafkaWrapper storage;
+
+    protected KafkaStorageFactory(Properties producerConfig,
```
Comment (dmytro-grankin) [inline kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStorageFactory.java]: This constructor is redundant.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import io.spine.server.storage.kafka.Topics.ValueTopic;
+
+import static io.spine.server.storage.kafka.Topics.PrefixedTopicFactory.FOR_AGGREGATE_RECORD;
+import static io.spine.server.storage.kafka.Topics.PrefixedTopicFactory.FOR_ENTITY_RECORD;
+import static io.spine.server.storage.kafka.Topics.PrefixedTopicFactory.FOR_EVENT_COUNT_AFTER_SNAPSHOT;
+import static io.spine.server.storage.kafka.Topics.PrefixedTopicFactory.FOR_LAST_EVENT_TIME;
+import static io.spine.server.storage.kafka.Topics.PrefixedTopicFactory.FOR_LIFECYCLE_FLAGS;
+
+/**
+ * An Apache Kafka topic.
```
Comment (dmytro-grankin) [inline kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java]: Please add some details about Kafka topics.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStandStorage.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        final StringValue key = checkNotNull(IdTransformer.direct().convert(id));
+        delegate.writeRecord(key, record);
+    }
+
+    @Override
+    protected void writeRecords(Map<AggregateStateId, EntityRecordWithColumns> records) {
+        final Map<StringValue, EntityRecordWithColumns> mapToWrite =
+                newHashMapWithExpectedSize(records.size());
+        final Function<AggregateStateId, StringValue> keyMapper = IdTransformer.direct()::convert;
+        for (Map.Entry<AggregateStateId, EntityRecordWithColumns> entry : records.entrySet()) {
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaStandStorage.java]: Left as is, as discussed vocally.

**Comment by dmdashenkov** [comment]
@dmytro-grankin, PTAL again.

**Pull request merged**